pkg/runtime/k0s/init.go

```diff
@@ -36,46 +42,100 @@ func (k *Runtime) init() error {
+	}
+	k.modifyConfigRepo()
+	if err := k.marshalToFile(); err != nil {
+		return fmt.Errorf("failed to write k0s config: %v", err)
 	}
-	ssh, err := k.getHostSSHClient(k.cluster.GetMaster0IP())
+	return nil
+}
+
+// GenerateCert generate the containerd CA for registry TLS and k0s join token for 100 years.
+func (k *Runtime) GenerateCert() error {
+	if err := k.generateK0sToken(); err != nil {
+		return err
+	}
+
+	if err := k.GenerateRegistryCert(); err != nil {
+		return err
+	}
+	return k.SendRegistryCert(k.cluster.GetMasterIPList()[:1])
+}
+
+func (k *Runtime) generateK0sConfig() error {
+	master0IP := k.cluster.GetMaster0IP()
+	ssh, err := k.getHostSSHClient(master0IP)
 	if err != nil {
-		return fmt.Errorf("failed to get ssh client: %v", err)
+		return err
 	}
-	output, err := ssh.Cmd(k.cluster.GetMaster0IP(), GetK0sVersionCMD)
+	createCMD := "mkdir -p /etc/k0s && k0s config create"
+	bytes, err := ssh.Cmd(master0IP, createCMD)
 	if err != nil {
 		return err
 	}
+	return yaml2.Unmarshal(bytes, &k.k0sConfig)
+}
 
-	k.K0sConfig.DefineConfigFork0s(string(output), k.RegConfig.Domain, k.RegConfig.Port, k.cluster.Name)
+func (k *Runtime) modifyConfigRepo() {
+	k.k0sConfig.Spec.Images.Repository = k.RegConfig.Domain + ":" + k.RegConfig.Port
+}
 
-	//write k0sctl.yaml to master0 rootfs
-	if err := k.K0sConfig.WriteConfigToMaster0(k.getRootfs()); err != nil {
+func (k *Runtime) marshalToFile() error {
+	bytes, err := yaml2.Marshal(k.k0sConfig)
+	if err != nil {
+		return err
+	}
+	if err = osi.NewAtomicWriter(DefaultK0sConfigPath).WriteFile(bytes); err != nil {
 		return err
 	}
 	return nil
 }
 
-// SSHKeyGenAndCopyIDToHosts use ssh-copy-id to prepare a no password ssh-client for k0sctl install environment.
-func (k *Runtime) SSHKeyGenAndCopyIDToHosts() error {
-	return k.sshKeyGenAndCopyIDToHosts()
+func (k *Runtime) BootstrapMaster0() error {
+	master0IP := k.cluster.GetMaster0IP()
+	ssh, err := k.getHostSSHClient(master0IP)
+	if err != nil {
+		return err
+	}
+	bootstrapCMD := fmt.Sprintf("k0s install controller -c %s", DefaultK0sConfigPath)
+	if _, err := ssh.Cmd(master0IP, bootstrapCMD); err != nil {
+		return err
+	}
+	startSvcCMD := "k0s start"
+	if _, err := ssh.Cmd(master0IP, startSvcCMD); err != nil {
+		return err
+	}
+	return nil
 }
 
-// GenerateCert generate the containerd CA for registry TLS.
-func (k *Runtime) GenerateCert() error {
-	if err := k.GenerateRegistryCert(); err != nil {
+func (k *Runtime) generateK0sToken() error {
+	master0IP := k.cluster.GetMaster0IP()
+	ssh, err := k.getHostSSHClient(master0IP)
+	if err != nil {
 		return err
 	}
-	return k.SendRegistryCert(k.cluster.GetMasterIPList()[:1])
+	workerTokenCMD := fmt.Sprintf("k0s token create --role=%s --expiry=876000h > %s", WorkerRole, DefaultK0sWorkerJoin)
+	if _, err := ssh.Cmd(master0IP, workerTokenCMD); err != nil {
+		return err
+	}
+	controllerTokenCMD := fmt.Sprintf("k0s token create --role=%s --expiry=876000h > %s", ControllerRole, DefaultK0sControllerJoin)
```

Comment:
use `CmdAsync` if we ignore the cmd output.

Reply:
Good suggestion! Thx